### PR TITLE
First Actor tutorial

### PR DESCRIPTION
**Pull request by jordan-rash**
https://www.katacoda.com/cjrash/courses/02-dev_apps/first-actor

### CODE REVIEW

**Comment by jordan-rash** [comment]
Goal: Highlight polyglot experience
Subgoal: shame ourselves into adding routes and token extraction to http actor interface

- [ ] Build a calculator actor
  - [ ] Provide add and divide, let user write subtract and multiple
- [ ] Create the actor’s logic via handlerequest
- [ ] Test invoke the actor via wash call
- [ ] Build, sign, run, and link actor in host (possibly via manifest as easiest)
  - [ ] or makefile? 
- [ ] Exercise the actor via curl and HTTP endpoint

**Comment by jordan-rash** [comment]
Suggested steps:
1. Describe goal of calculator app
  - Have all three languages so user can pick adventure
2. explain how add and divide work (with respect to wasmcloud)
3. guide user though subtract and multiple
4. invoke multiply in wash
5. describe different ways to run (allow user to pick)
  - manifest
  - within wash
6.  curl actor from commandline

File: dev_apps/first-actor/buildsign.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+First off, lets build and sign the actor as-is.  This way, we can see if the module is behaving as we expect.
+
+
+<details>
+  <summary>Rust</summary>
+  Build it!
+  `cargo build --release`{{execute}}
+  Sign it!
+  `wash claims sign target/wasm32-unknown-unknown/release/calculator.wasm -c wasmcloud:httpserver --name "calculator" --ver 0.1.0 --rev 0`
```
Comment (autodidaddict) [inline dev_apps/first-actor/buildsign.md:11]: the wasmcloud:httpserver is a first-class (well known) claim name. We should avoid using `-c` for anything other than custom (non-first-party) claims. The proper switch for `wasmcloud:httpserver` is `-q`

File: dev_apps/first-actor/buildsign.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  Sign it!
+  `wash claims sign target/wasm32-unknown-unknown/release/calculator.wasm -c wasmcloud:httpserver --name "calculator" --ver 0.1.0 --rev 0`  
+</details>
+<details>
+  <summary>Go</summary>
+
+  Build it!
+  `mkdir -p build && tinygo build -o build/calculator.wasm -target wasm -no-debug main.go`{{execute}}
+  Sign it!
+  `wash claims sign build/calculator.wasm -c wasmcloud:httpserver --name "calculator" --ver 0.1.0 --rev 0`{{execute}}
```
Comment (autodidaddict) [inline dev_apps/first-actor/buildsign.md:19]: Ditto, use `-q` instead of `-c` here.

File: dev_apps/first-actor/buildsign.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+</details>
+<details>
+  <summary>Assembly Script</summary>
```
Comment (autodidaddict) [inline dev_apps/first-actor/buildsign.md:22]: `AssemblyScript` is written as one word.

File: dev_apps/first-actor/buildsign.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  
+  Build it!
+  ????
+  Sign it!
+  ????
+</details>
+
+> Notice: We have signed the `wasmcloud` module with the `-c wasmcloud:httpserver` flag.  This essentially gives our actor the permissions it will need to communicate with a `httpserver` provider when the time comes.
+
+Once we run `wash claims sign`, on a wasm module, notice that a new file is dropped with the same name as our original file, with a `-s` appended to it.  This is a sign `wasmcloud` ACTOR!!!  At this point, we can use was to checkout the metadata of our branch new actor!
```
Comment (autodidaddict) [inline dev_apps/first-actor/buildsign.md:32]: It's a `_s` that's appended, not `-s`. 

> This is a sign `wasmcloud` ACTOR!!!
replace:
This is a signed wasmCloud actor.

> At this point, we can use was to checkout the metadata of our branch new actor!
replace:
At this point, we can use wash to examine the metadata of our brand new actor.

File: dev_apps/first-actor/buildsign.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+<details>
+  <summary>Rust</summary>
+  Inspect it!
```
Comment (autodidaddict) [inline dev_apps/first-actor/buildsign.md:36]: Avoid using exclamation points unless absolutely necessary.

File: dev_apps/first-actor/choose.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Choose your adventure
+
+To demonstrate how `wasmcloud` leverages the powerful polygot nature of webassembly, we have provided you with three identical calculators, written in three different languages.  Throughout the tutorial, you will see sections similiar to this one:
```
Comment (autodidaddict) [inline dev_apps/first-actor/choose.md:3]: don't use `wasmcloud`, instead just "wasmCloud" as a word. 
**WebAssembly** is always written with a capital W and capital A.

typo: spelled polyglot wrong.

File: dev_apps/first-actor/choose.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+</details>
+<details>
+  <summary>Assembly Script</summary>
+
+  `cd assemblyscript`{{execute}}  
+</details>
+
+Make sure that if you use the "click & execute" capabilities of Katacoda, you stick with the same language drop downs all the way through the tutorial.  
+
+Awesome!  Lets get started....open a language above and `cd` into the directory
```
Comment (autodidaddict) [inline dev_apps/first-actor/choose.md:23]: s/open/select

File: dev_apps/first-actor/index.json [hunk 8]
```diff
@@ -1,24 +1,33 @@
 {
-  "title": "Wash101 - First Actor",
-  "description": "Getting familiar with the actor workflow",
+  "title": "Your First Actor",
+  "description": "Walk through editing, signing, deploying, and calling a calculator actor in multiple languages.",
```
Comment (autodidaddict) [inline dev_apps/first-actor/index.json:3]: Edit, sign, deploy, and call a calculator actor in multiple languages

File: dev_apps/first-actor/intro.md [hunk 9]
```diff
@@ -1,4 +1,12 @@
-## wash101
+# Your First wasmcloud Actor!
```
Comment (autodidaddict) [inline dev_apps/first-actor/intro.md:1]: # Your First wasmCloud Actor

File: dev_apps/first-actor/intro.md [hunk 10]
```diff
@@ -1,4 +1,12 @@
...
 
-## Your first actor
-In this tutorial, you will deploy your first actor using the wash REPL
+Congrats on making it this far!  Now it is time to take the take what we have learn from the basics courses and build our first actor.  How about a calculator?!
```
Comment (autodidaddict) [inline dev_apps/first-actor/intro.md:3]: _How about a calculator?_

File: dev_apps/first-actor/intro.md [hunk 11]
```diff
@@ -1,4 +1,12 @@
...
+
+In this tutorial, we have provided you with 3 implementations of the same calculator app.  They are written in Rust, Golang, and AssemblyScript.  You can choose your own adventure, or you can do them all!  One of the tenants of wasmcloud is that it allows you to build on the polyglot power of Webassembly.
```
Comment (autodidaddict) [inline dev_apps/first-actor/intro.md:5]: written in Rust, Go, and AssemblyScript

by convention, we don't use `golang` unless in metadata/SEO and search requests. The name of the language is Go.

Comment (autodidaddict) [reply dev_apps/first-actor/intro.md:5]: s/tenants/tenets
s/wasmcloud/wasmCloud

Comment (autodidaddict) [reply dev_apps/first-actor/intro.md:5]: > Webassembly
WebAssembly

File: dev_apps/first-actor/intro.md [hunk 12]
```diff
@@ -1,4 +1,12 @@
...
+# Your First wasmcloud Actor!
 
-## Your first actor
-In this tutorial, you will deploy your first actor using the wash REPL
+Congrats on making it this far!  Now it is time to take the take what we have learn from the basics courses and build our first actor.  How about a calculator?!
+
+In this tutorial, we have provided you with 3 implementations of the same calculator app.  They are written in Rust, Golang, and AssemblyScript.  You can choose your own adventure, or you can do them all!  One of the tenants of wasmcloud is that it allows you to build on the polyglot power of Webassembly.  
+
+By the end of this tutorial you should have an understanding of:
+- How wasmcloud leverages the polyglot power of webassembly
```
Comment (autodidaddict) [inline dev_apps/first-actor/intro.md:8]: wasmCloud
WebAssembly

File: dev_apps/first-actor/intro.md [hunk 13]
```diff
@@ -1,4 +1,12 @@
...
+- Signing a wasm module with `wash`
+- Pushing a signed `wasmcloud` module to an OCI compliant registry
```
Comment (autodidaddict) [inline dev_apps/first-actor/intro.md:10]: no backticks on wasmcloud. When we say "a signed wasmCloud module" it's referring to wasmCloud as a non, not as the name of a binary file run at a command line.

**Review by autodidaddict — CHANGES_REQUESTED** [review]
Some nitpicks and copy editing

File: dev_apps/first-actor/buildsign.md (hunk 1 above)
Comment (brooksmtownsend) [reply dev_apps/first-actor/buildsign.md:11]: I would suggest using `--http_server` for clarity in the intro tutorial

File: dev_apps/first-actor/buildsign.md (hunk 2 above)
Comment (brooksmtownsend) [reply dev_apps/first-actor/buildsign.md:19]: I would suggest using --http_server for clarity in the intro tutorial

**Review by autodidaddict — APPROVED** [review]

File: dev_apps/first-actor/buildsign.md (hunk 5 above)
Comment (jordan-rash) [reply dev_apps/first-actor/buildsign.md]: Listen, at my age....21.....exclamation points are an excepted practice.  I think it attracts the young, vibrant part of the community

File: dev_apps/first-actor/intro.md (hunk 9 above)
Comment (jordan-rash) [reply dev_apps/first-actor/intro.md]: at least I didnt use ticks there....#progress

File: 02-dev_apps/first-actor/buildsign.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+</details>
+<details>
+  <summary>AssemblyScript</summary>
+  
+  Build it    
+  `npm run build`{{execute}}
+
+Sign it  
+ `wash claims sign calculator.wasm --http_server --name "calculator" --ver 0.1.0 --rev 0`{{execute}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/buildsign.md:31]: for some reason "sign it" shows up on the same line as the build it script, is that due to indents? (not a big deal)

File: 02-dev_apps/first-actor/buildsign.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+</details>
+
+> Notice: We have signed the wasmCloud module with the `--http_server` flag. This essentially gives our actor the permissions it will need to communicate with a `httpserver` provider when the time comes.
+
+Once we run `wash claims sign`, on a wasm module, notice that a new file is dropped with the same name as our original file, with a `_s` appended to it. This is a sign wasmCloud actor. At this point, we can use wash to examine the metadata of our branch new actor.
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/buildsign.md:37]: `s/sign/signed`

File: 02-dev_apps/first-actor/buildsign.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  <summary>Go</summary>
+
+`wash reg push localhost:5000/calc:0.1.0 build/calculator_s.wasm --insecure`{{execute}}
+
+</details>
+<details>
+  <summary>Assembly Script</summary>
+  
+`wash reg push localhost:5000/calc:0.1.0 calculator_s.wasm --insecure`{{execute}}
+</details>
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/buildsign.md:79]: Same audible as the table_tennis, let's just skip the registry (feel free to remove the registry from the docker image too if that speeds things up)

File: 02-dev_apps/first-actor/intro.md [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+
+Congrats on making it this far! Now it is time to take the take what we have learn from the basics courses and build our first actor. _How about a calculator?_
+
+In this tutorial, we have provided you with 3 implementations of the same calculator app. They are written in Rust, Go, and AssemblyScript. You can choose your own adventure, or you can do them all! One of the tenets of wasmCloud is that it allows you to build on the polyglot power of WebAssembly.
+
+By the end of this tutorial you should have an understanding of:
+
+- How wasmCloud leverages the polyglot power of WebAssembly
+- Signing a wasm module with `wash`
+- Pushing a signed wasmCloud module to an OCI compliant registry
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/intro.md:11]: I think we should remove the OCI compliant registry part

File: 02-dev_apps/first-actor/manifest.md [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+`assemblyscript/manifest.yaml`{{open}}
+
+</details>
+
+It is important to notice that `actor` fields in the link has an ID provided. That ID will NOT be the same for you on your system. We have two options:
+
+- You can use `wash` to give you the actor ID of your actor
+  `wash claims inspect localhost:5000/calc:0.1.0 --insecure`{{execute}}
+- You can use the environmental variable to override the current ID
+  `export CLIENT_ACTOR=$(wash claims inspect localhost:5000/calc:0.1.0 --insecure -o json | jq -r '.module')`{{execute}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/manifest.md:27]: change this to `wash claims inspect` with the signed wasm module (you may need to put this in a polyglot section, apologies)

File: 02-dev_apps/first-actor/manifest.md [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+Once you have the `manifest.yml` updated with the correct `actor_id`, you can start the run time and interface with your actor!
+
+> Note: At this point, everything is in place to test the current state of the actor!
+> `wasmcloud --allowed-insecure localhost:5000 -m manifest.yaml`{{execute}} > `curl localhost:8080/add?5,5 --output -`{{execute T2}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/manifest.md:32]: After getting rid of the registry you can remove `--allowed-insecure localhost:5000`

File: 02-dev_apps/first-actor/run.md [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+Lets build it, sign it, and push it again.
+
+<details>
+  <summary>Rust</summary>
+  
+  `cargo build --release`{{execute interrupt T1}}   
+  
+  `wash claims sign target/wasm32-unknown-unknown/release/calculator.wasm -c wasmcloud:httpserver --name "calculator" --ver 0.1.0 --rev 0`{{execute T1}}
+
+`wash drain all && wash reg push localhost:5000/calc:0.1.0 target/wasm32-unknown-unknown/release/calculator_s.wasm --insecure`{{execute T1}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/run.md:12]: after registry change, you can remove this

File: 02-dev_apps/first-actor/run.md [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+
+</details>
+<details>
+  <summary>Go</summary>
+
+`mkdir -p build && tinygo build -o build/calculator.wasm -target wasm -no-debug main.go`{{execute interrupt T1}}
+
+`wash claims sign build/calculator.wasm -c wasmcloud:httpserver --name "calculator" --ver 0.1.0 --rev 0`{{execute T1}}
+
+`wash drain all && wash reg push localhost:5000/calc:0.1.0 build/calculator_s.wasm --insecure`{{execute T1}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/run.md:22]: after registry change, you can remove this

File: 02-dev_apps/first-actor/run.md [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+</details>
+<details>
+  <summary>AssemblyScript</summary>
+
+`npm run build`{{execute interrupt T1}}
+
+`wash claims sign calculator.wasm -q --name "calculator" --ver 0.1.0 --rev 0`{{execute T1}}
+
+`wash drain all && wash reg push localhost:5000/calc:0.1.0 calculator_s.wasm --insecure`{{execute T1}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/run.md:32]: after registry change, you can remove this

File: 02-dev_apps/first-actor/run.md [hunk 23]
```diff
@@ -0,0 +1,45 @@
+We have come to the end of our journey. We have a `wasmcloud` module, an updated manifest file, and a massive level of excitement! Here we go!
+
+Lets build it, sign it, and push it again.
+
+<details>
+  <summary>Rust</summary>
+  
+  `cargo build --release`{{execute interrupt T1}}   
+  
+  `wash claims sign target/wasm32-unknown-unknown/release/calculator.wasm -c wasmcloud:httpserver --name "calculator" --ver 0.1.0 --rev 0`{{execute T1}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/run.md:10]: let's change `-c wasmcloud:httpserver` to `--http_server`

File: 02-dev_apps/first-actor/run.md [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+
+`wash drain all && wash reg push localhost:5000/calc:0.1.0 target/wasm32-unknown-unknown/release/calculator_s.wasm --insecure`{{execute T1}}
+
+</details>
+<details>
+  <summary>Go</summary>
+
+`mkdir -p build && tinygo build -o build/calculator.wasm -target wasm -no-debug main.go`{{execute interrupt T1}}
+
+`wash claims sign build/calculator.wasm -c wasmcloud:httpserver --name "calculator" --ver 0.1.0 --rev 0`{{execute T1}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/run.md:20]: let's change `-c wasmcloud:httpserver` to `--http_server`

File: 02-dev_apps/first-actor/run.md [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+
+`wash drain all && wash reg push localhost:5000/calc:0.1.0 build/calculator_s.wasm --insecure`{{execute T1}}
+
+</details>
+<details>
+  <summary>AssemblyScript</summary>
+
+`npm run build`{{execute interrupt T1}}
+
+`wash claims sign calculator.wasm -q --name "calculator" --ver 0.1.0 --rev 0`{{execute T1}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/run.md:30]: let's change `-q` to `--http_server`

File: 02-dev_apps/first-actor/run.md [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+
+`wash drain all && wash reg push localhost:5000/calc:0.1.0 calculator_s.wasm --insecure`{{execute T1}}
+
+</details>
+
+> Note: We prefaced the registry push with a `wash drain all` which cleared the OCI cache. We did this since we are reusing the same tag.
+
+And now lets start `wasmcloud` with our `manifest.yaml`!
+`wasmcloud --allowed-insecure localhost:5000 -m manifest.yaml`{{execute interrupt T1}}
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/run.md:39]: we can remove `--allowed-insecure localhost:5000`

File: 02-dev_apps/first-actor/edit.md [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+
+<details>
+  <summary>Rust</summary>
+
+  <pre class="file" data-filename="rust/src/lib.rs" data-target="insert" data-marker="// TODO: add multiplication">
+
+Open the file:
+
+`rust/src/lib.rs`{{open}}  
+
```
Comment (brooksmtownsend) [inline 02-dev_apps/first-actor/edit.md:11]:
```
Open the file:

`rust/src/lib.rs`{{open}} 
```
this section is also copied to the editor, I think it should be outside the `<pre>` tag

**Review by brooksmtownsend — DISMISSED** [review]
A couple changes, mostly relating to removing the registry

**Pull request merged**